### ISSUE TITLE
IsEquivalentTo wrongly fails when comparing dictonary-of-dictonaries, that are identical except the order of keys in child dicts

### ISSUE THREAD

**Issue by tpehkone** [tpierrain/NFluent#307]
## Expected Behavior

When comparing structures with nested dictionaries, the order of keys of the nested dictionaries should not matter.

## Current Behavior

IsEquivalentTo seems to allow any order of keys only if comparing trivial one level dictionaries. But for nested dictionaries the order matters, possibly making the comparison to fail.

The same happens, if trying to use: Contains() , Contains().Only() , IsEqualTo() . In fact, I was unable to find out any operation that would successfully compare nested dictionaries.

Furthermore, it looks like also lists-of-dictionaries suffer of the same issue, but did not yet test them thoroughly.

## Code that reproduce the issue 

```csharp
[Test]
public void Test_DictOfDict_IsEquivalentTo()
{
    var dictOf3_A = new Dictionary<string, string> { { "aa", "AA" }, { "bb", "BB" }, { "cc", "CC" } };
    var dictOf3_B = new Dictionary<string, string> { { "cc", "CC" }, { "aa", "AA" }, { "bb", "BB" } };
    var dictOf2 = new Dictionary<string, string> { { "cc", "CC" }, { "bb", "BB" } };

    var dictOf2Dict_A = new Dictionary<string, Dictionary<string, string>> { { "key1", dictOf2 }, { "key2", dictOf3_A } };
    var dictOf2Dict_B = new Dictionary<string, Dictionary<string, string>> { { "key1", dictOf2 }, { "key2", dictOf3_B } };

    Check.That(dictOf2Dict_A).IsEquivalentTo(dictOf2Dict_B);  // Fails -> Too picky on evaluating child-dict order
}
```

## Your Environment

* Version used: 2.6.0
* Framework version: .Net Core 2.1

**Comment by dupdob** [comment]
I published a V2.6.1 beta that fixes this issue.
Can you please try it and confirm issue is solved?
Package is available on MyGet:
https://www.myget.org/feed/dupdobnightly/package/nuget/NFluent/2.6.1-beta-0204

**Comment by tpehkone** [comment]
Thanks for the prompt first fix on this issue. Dict-of-dict structure comparisons seem to work fine now, so maybe this issue can be considered closed.

As a heads-up however, I need to do some further testing to exactly pinpoint some remaining problematic ares, possibly raising up separate issues then - but that goes for the next week. Some initial finding with the beta:

* Contains() and Contains().Only() structures have similar key order issue than described here, for dict-of-dicts.
* **IsEquivalentTo _might_ still have key issue with params-of-lists-of-dicts (have to reproduce minimun case to confirm that.)**
* I agree there is potential need for both fully weak and fully strong type equality check, 
but for most of my own testing purposes I would still need some mechanism for strong-list-equality && weak-dict-equality check realized by one top level algorithm.
* Not tested yet with generic types of IEnumerable<T> / IEnumerable<KeyValuePair<Tkey, Tvalue>> / ICollection , but will get there eventually, too.

Background: I have a rather extensive self-test set against an automatic-data-conversion-layer between 2 technologies, supporting conversion of for all kinds of array/list/dict nested constructs. NFluent comes into play now that we are trying to find a new, self-standing and clean assertion library to do all the heavy lifting of data comparison work.

**Comment by dupdob** [comment]
Thanks for providing me with all these details. I will gladly work with you regarding your use cases. I am looking for this kind of feedback to help me drive the evolution of the lib.
Coming back to your comments:
1. Contains... relies on strict equality, so this is not surprising.  To be more specific, it relies on the default equality strategy, which can select using [**Check.EqualMode**](https://github.com/tpierrain/NFluent/wiki/Configuring-NFluent#equalmode). So I can add a **WeakEqual** option (naming TBC) that would disregard the order
2. IsEquivalent still has issues with dict of lists of dicts:  I think it should not be the case; I will try to reproduce on my own if time permits but any sample you could provide would be helping of course.
3. By params I assume you mean variable list of parameters, as in ```Check.That(dico).IsEquivalentTo(list1, list2, list3v ...)```. It may probably be linked to the previous issue.
4. As discussed above, this could be controlled using the static parameter
5. Usage of nested generics will probably get disappointing result, unless they implement IDictionary. The fact is nesting requires type erasure for recursion, the alternative being to declare all possible signatures, including multi-level nesting.

Also, if you are interested, we have a [Gitter](https://gitter.im/nfluent/community) channel and a Slack for more interactive exchanges.

**Comment by vivainio** [comment]
- I'm not sure order-agnostic dictionary comparison should be considered "weak". Normal structural comparison in languages that support it (python, F#...) ignores insertion order in dicts and sets

- Deep comparison using "object" and reflection would probably address the deep comparison use case raised by @tpehkone

**Comment by dupdob** [comment]
NFluent V 2.6.2 beta 205 is available on [MyGet](https://www.myget.org/feed/dupdobnightly/package/nuget/NFluent/2.6.1-beta-0205).
It fixes the problem with **IsEquivalent** for _List_ and other _IEnumerables_ (except arrays).

@tpehkone: I would value your feedback on this version. My plan is to:
1. implements what should be the proper behaviour for **IsEquivalent** for the major type
2. extend to more exotic types (such as custom IDictionary<K,V> implementations)
3. refactor to good design
4. make a public release

**Comment by dupdob** [comment]
fixed in V 2.7.0
